src/dawn/CodeGen/ASTCodeGenGTClangStencilDesc.cpp

```diff
@@ -62,12 +87,103 @@ void ASTCodeGenGTClangStencilDesc::visit(const std::shared_ptr<VerticalRegionDec
 
 void ASTCodeGenGTClangStencilDesc::visit(const std::shared_ptr<StencilCallDeclStmt>& stmt) {
   int StencilID = instantiation_->getStencilCallToStencilIDMap().find(stmt)->second;
+  StencilIDsVisited_.push_back(StencilID);
   for(const std::string& stencilName : StencilIDToStencilNameMap_.find(StencilID)->second)
     ss_ << std::string(indent_, ' ') << stencilName << ".get_stencil()->run();\n";
 }
 
 void ASTCodeGenGTClangStencilDesc::visit(const std::shared_ptr<BoundaryConditionDeclStmt>& stmt) {
-  DAWN_ASSERT_MSG(0, "BoundaryConditionDeclStmt not yet implemented");
+
+  auto calculateHaloExtents = [&](std::string fieldname) {
+
+    Extents fullExtent;
+    // Did we already apply a BoundaryCondition for this field?
+    // This is the first time we apply a BC to this field, we traverse all stencils that were
+    // applied before
+    if(StencilBCsApplied_.count(fieldname) == 0) {
```

Comment:
I think the if and the else are pretty similar, so I would try to merge the algorithm, by first extracting the
`stencilIDsToVisit` which in the case of the if is just a copy of `StencilIDsVisited_`, and from this on the algorithm should be the same

Reply:
true, fixed